### ISSUE TITLE
Pyright does not detect difference in Protocol and Inherited method signature.

### ISSUE THREAD

**Issue by rshanor** [microsoft/pyright#10183]
**Describe the bug**

- We have a protocol P with some method signature `M(self, str)`
- We have a class C which inherits from P
- C has a method on it with a signature `M(self, Optional[str])`
- Pyright fails to detect that `P.M` != `C.M`

**Code or Screenshots**

```python
from typing import Optional, Protocol

class Iface(Protocol):
    def do_something(self, msg: str) -> None: ...

class ThisShouldNotTypeCheckImpl(Iface):
    def do_something(self, msg: Optional[str]) -> None:
        print(msg)

def main() -> None:
    optional = ThisShouldNotTypeCheckImpl()
    optional.do_something(None)
```

**VS Code extension or command-line**
I am using `pyright 1.1.397`

**Comment by erictraut** [comment]
Pyright's behavior here is correct. Not surprisingly, mypy also accepts the above code.

The signature of a method in an implementation doesn't need to match exactly for a protocol match. The implementation's method must be _assignable to_ the protocol's method. Refer to the [assignability rules for callable types](https://typing.python.org/en/latest/spec/callables.html#callables) in the typing spec for details.

In your example, the protocol (which is a structural type definition) indicates that the `do_something` method must accept a single argument of type `str`. The class `ThisShouldNotTypeCheckImpl` provides an implementation that meets this requirement. The fact that the class supports additional functionality (namely, the ability to accept a `None` argument) does not make it incompatible with the protocol's contract.

**Comment by rshanor** [comment]
@erictraut ahh I see... Thanks for the explanation, makes sense.
